### PR TITLE
Match query to respect lenient flag

### PR DESCRIPTION
**Pull request by javanna**
A recent change (#63406) introduced throwing exception when a match query is executed against a field that does not support text search. The query would fail anyways, only at a later stage. The problem though is that the lenient flag needs to be taken into account.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/search/MatchQuery.java [hunk 1]
```diff
@@ -297,13 +307,7 @@ protected final Query parseInternal(Type type, String fieldName, MatchQueryBuild
     }
 
     protected Analyzer getAnalyzer(MappedFieldType fieldType, boolean quoted) {
-        // We check here that the field supports text searches and therefore has an analyzer -
-        // if it doesn't, we can bail out early without doing any further parsing.
         TextSearchInfo tsi = fieldType.getTextSearchInfo();
```
Comment (romseygeek) [inline server/src/main/java/org/elasticsearch/index/search/MatchQuery.java:310]: Maybe assert that this != TextSearchInfo.NONE?

File: server/src/test/java/org/elasticsearch/index/query/MatchQueryBuilderTests.java [hunk 2]
```diff
@@ -297,7 +297,16 @@ public void testExactOnUnsupportedField() throws Exception {
...
     }
 
-    public void testParseFailsWithMultipleFields() throws IOException {
+    public void testLenientFlag() throws Exception {
+        MatchQueryBuilder query = new MatchQueryBuilder(BINARY_FIELD_NAME, "test");
+        QueryShardContext context = createShardContext();
+        IllegalArgumentException e = expectThrows(IllegalArgumentException.class, () -> query.toQuery(context));
+        assertEquals("Field [mapped_binary] of type [binary does not support match queries", e.getMessage());
+        query.lenient(true);
+        query.toQuery(context); // no exception
```
Comment (romseygeek) [inline server/src/test/java/org/elasticsearch/index/query/MatchQueryBuilderTests.java:306]: check that this is a MatchNoDocsQuery?

**Review by romseygeek — APPROVED** [review]
LGTM, I left a couple of suggestions to improve testing/assertions

**Comment by javanna** [comment]
run elasticsearch-ci/1

**Pull request merged**